contracts/eosio.system/src/delegate_bandwidth.cpp

```diff
@@ -206,17 +252,26 @@ namespace eosiosystem {
       require_recipient( owner );
    }
 
+   // this is called when transfering or selling ram, and encumbered (gifted) ram cannot be sold or transferred.
+   // so if we hold some gifted ram, deduct the amount from the ram available to be sold or transfered.
    int64_t system_contract::reduce_ram( const name& owner, int64_t bytes ) {
       check( bytes > 0, "cannot reduce negative byte" );
+
       user_resources_table userres( get_self(), owner.value );
       auto res_itr = userres.find( owner.value );
       check( res_itr != userres.end(), "no resource row" );
-      check( res_itr->ram_bytes >= bytes, "insufficient quota" );
+
+      gifted_ram_table giftedram( get_self(), get_self().value );
+      auto giftedram_itr = giftedram.find( owner.value );
+      bool holding_gifted_ram = (giftedram_itr != giftedram.end());
+
+      auto available_bytes = holding_gifted_ram ? res_itr->ram_bytes - giftedram_itr->ram_bytes : res_itr->ram_bytes;
+      check( available_bytes >= bytes, "insufficient quota" );
```

Comment:
Could improve this error message to include the existing available_bytes.

Reply:
This is used by other APIs like `ramtransfer`, `ramburn`, `sellram`, etc... and would require changes in tests I didn't write (which is fine), but I wonder if someone else could rely on the exact error message.

Reply:
Up to you and @arhag. Not a big deal.

Reply:
I think it's okay to keep this error message as is.